recommence/Checkpoint.py

```diff
@@ -0,0 +1,38 @@
+import os
+import shutil
+import pickle
+from typing import Any, Dict, Callable
+
+class Checkpoint:
+    def __init__(self, save_path: str):
+        self.save_path: str = save_path
+        self._data_path: str = f'{save_path}/data.pkl'
+        self._data: Dict[str, Any] = {}
+
+        self._load_if_exists()
+
+    def __getitem__(self, name: str) -> Any:
+        return self._data[name]
+
+    def __setitem__(self, name: str, value: Callable[[], Any]) -> Any:
```

Comment:
Not an issue for this PR, just thinking about some next steps:

We should aim to have some usage pattern that doesn't cause type widening. For instance, I'd love if the type-checker knew that the variable `agent` below is of type `Agent`. Instead, it is currently inferred as `Any`.
```python
chk['agent'] = lambda: Agent(...)
agent = chk['agent']
```

Brainstorming some strategies. Could make the api look something like:
```python
agent = chk.register('agent', lambda: Agent(...))
```
similar to the older api. I'm remembering, now, why I had designed it that way previously.